File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int solve(TreeNode* root, int& maxi){
        if(!root) return 0;
        int lh = max(solve(root->left, maxi),0);
        int rh = max(solve(root->right, maxi),0);
        maxi = max(maxi, root->val + lh + rh);
        return root->val + max(rh,lh);
    }
    int maxPathSum(TreeNode* root) {
        int maxi = root->val;
        solve(root, maxi);
        return maxi;
    }
};

File: Data Structures & Algorithms/house-robber-iii/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        dp = {}
        def solve(root):
            if not root:
                return 0
            if root in dp:
                return dp[root]
            take = root.val
            if root.left:
                take += solve(root.left.left)
                take += solve(root.left.right)
            if root.right:
                take += solve(root.right.left)
                take += solve(root.right.right)
            notake = 0
            if root.left:
                notake += solve(root.left)
            if root.right:
                notake += solve(root.right)
            dp[root] = max(take, notake)
            return dp[root]
        return solve(root)

            